src/Microsoft.Data.Analysis/DataFrame.IO.cs

```diff
@@ -198,55 +198,68 @@ private static DataFrameColumn CreateColumn(Type kind, string[] columnNames, int
                                 Encoding encoding = null)
         {
             if (!csvStream.CanSeek)
+            {
                 throw new ArgumentException(Strings.NonSeekableStream, nameof(csvStream));
+            }
+
+            if (dataTypes == null && guessRows <= 0)
+            {
+                throw new ArgumentException(string.Format(Strings.ExpectedEitherGuessRowsOrDataTypes, nameof(guessRows), nameof(dataTypes)));
+            }
 
             var linesForGuessType = new List<string[]>();
             long rowline = 0;
             int numberOfColumns = dataTypes?.Length ?? 0;
 
             if (header == true && numberOfRowsToRead != -1)
+            {
                 numberOfRowsToRead++;
+            }
 
             List<DataFrameColumn> columns;
             long streamStart = csvStream.Position;
             // First pass: schema and number of rows.
             using (var streamReader = new StreamReader(csvStream, encoding ?? Encoding.UTF8, detectEncodingFromByteOrderMarks: true, DefaultStreamReaderBufferSize, leaveOpen: true))
             {
                 string line = null;
-                if (dataTypes == null)
+                line = streamReader.ReadLine();
+                while (line != null)
                 {
-                    line = streamReader.ReadLine();
-                    while (line != null)
+                    if ((numberOfRowsToRead == -1) || rowline < numberOfRowsToRead)
                     {
-                        if ((numberOfRowsToRead == -1) || rowline < numberOfRowsToRead)
+                        if (linesForGuessType.Count < guessRows || guessRows == 0)
```

Comment:
Isn't the right check here:

`if (linesForGuessType.Count < guessRows || (header && rowline == 0))`

?

If `guessRows == 0` and `!header`, then there is no reason to get here, right?

Reply:
Yup, your idea is better. The only other thing I had to do was to change 
``` csharp
 if (rowline == guessRows)
                    {
                        break;
                    }
```
to 
``` csharp
 if (rowline == guessRows || guessRows == 0)
                    {
                        break;
                    }
```
Otherwise we'd end up reading the whole stream when guessRows = 0 unnecessarily.